lib/authorize.js

```diff
@@ -0,0 +1,68 @@
+const { STATUS_CODES } = require('../constants/constants');
+const { getResourcePermissions } = require('../src/modules/resources/service/assignResourcePermission');
+const { getResourceData } = require('../src/modules/resources/service/getResource');
+const { getRolePermissions } = require('../src/modules/roles/service/assignRolePermissionService');
+const {
+  getUserRoles,
+} = require('../src/modules/user/service/userRoleService');
+const { getUserPermissions } = require('../src/modules/user/service/userService');
+const redisClient = require('../utils/redisClient');
+
+const authorize = (resource, requiredPermissions) => {
+  return async (req, res, next) => {
+    try {
+      const { userId } = req.user;
+
+      const cachedUser = await redisClient.get(`user:${userId}`);
+
+      let userData;
+      if (cachedUser) {
+        userData = JSON.parse(cachedUser);
+      } else {
+        // Fallback to DB if not in Redis
+        const roles = await getUserRoles(userId);
```

Comment:
in this case we should store the permissions in redis as well so that in next iteration it does not gets missed being present in cache